File: ios/Classes/ThunderIDFlutterPlugin.swift
import Flutter
import UIKit
import ThunderID

@objc public class ThunderIDFlutterPlugin: NSObject, FlutterPlugin {
    private let handler = ThunderIDMethodHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "dev.thunderid/sdk", binaryMessenger: registrar.messenger())
        let instance = ThunderIDFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        Task { @MainActor [handler] in
            await handler.handle(method: call.method, args: args, result: result)
        }
    }
}

File: android/src/main/kotlin/dev/thunderid/flutter/ThunderIDFlutterPlugin.kt
package dev.thunderid.flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class ThunderIDFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var handler: ThunderIDMethodHandler
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        handler = ThunderIDMethodHandler(binding.applicationContext)
        channel = MethodChannel(binding.binaryMessenger, "dev.thunderid/sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val args = call.arguments<Map<String, Any?>>() ?: emptyMap()
        scope.launch {
            handler.handle(call.method, args, result)
        }
    }
}
